# 0828

## 예외처리(Exception handling)

### 프로그램 오류

#### 컴파일 에러(compile-time error)와 런타임 에러(runtime error)

- 컴파일 에러 – 컴파일할 때 발생하는 에러
- 런타임 에러 – 실행할 때 발생하는 에러

#### Java의 런타임 에러 – 에러(error)와 예외(exception)
 
```
에러(error) - 프로그램 코드에 의해서 수습될 수 없는 심각한 오류
예외(exception) - 프로그램 코드에 의해서 수습될 수 있는 다소 미약한 오류
```

### 예외처리의 정의와 목적

#### 에러(error)는 어쩔 수 없지만, 예외(exception)는 처리해야 한다

```
에러(error) - 프로그램 코드에 의해서 수습될 수 없는 심각한 오류
예외(exception) - 프로그램 코드에 의해서 수습될 수 있는 다소 미약한 오류
```

#### 예외처리의 정의와 목적

```
예외처리(exception handlng)
정의 - 프로그샘 실행 시 발생할 수 있는 예외의 발생에 대비한 코드를 작성하는 것
목적 - 프로그램의 비정상 종료를 막고, 정상적인 실행상태를 유지하는 것
```

### 예외처리구문 - try - catch

```
try{
  // 예외가 여기 발생할 가능성이 있는 문장들을 넣기
} catch (Exctption e1) {
  // Exception1이 발생했을 경우, 이를 처리하기 위한 문장 작성
} catch (Exctption e2) {
  // Exception2이 발생했을 경우, 이를 처리하기 위한 문장 작성
} ....
catch (Exctption eN) {
  // ExceptionN이 발생했을 경우, 이를 처리하기 위한 문장 작성
}
```

### try-catch문에서의 흐름

- try블럭 내에서 예외가 발생한 경우.

1. 발생한 예외와 일치하는 catch블럭이 있는지 확인
2. 일치하는 cathc블럭을 찾게 되면, 그 catch블럭 내의 문장들을 수행하고 전체 try-cathc문을 빠져나가서 그 다음 문장을 계속해서 수행한다. 만일 일치하는 catch블럭을 찾지 못 하면, 예외는 처리되지 못한다.

- try블럭 내에서 예외가 발생하지 않은 경우.

1. catch블럭을 거치지 않고 전체 try-catch문을 빠져나가서 수행을 계속한다.


### 예외 발생시키기

1. 먼저 연산자 new를 이용해서 발생시키려는 예외 클래스의 객체를 만든 다음
2. 키워드 throw를 이용해서 예외를 발생시킨다.

```
Exception e = new Exception("고의로 발생"); // (1)
thorw e; // (2)
```

### 예외 클래스의 계층구조

- 예외 클래스는 크게 두 그룹으로 나뉜다

```
RuntimeException 클래스들 - 프로그래머의 실수로 발생하는 예외 <- 예외처리 필수
Exception 클래스들 - 사용자의 실수와 같은 외적인 요인에 의해 발생하는 예외 <- 예외처리 선택
```

레퍼런스가 null인 변수 클래스들은 메서드 호출 시 에러 발생!

자식 예외를 부모 캐치가 받을 수 있다

예외는 대체처리

잘 만드는 예외처리는 각각 예외마다 디테일 하게 처리하는 것

상속관계가 큰 catch는 가장 아래로 보낸다, 즉 최상의 예외 Exception는 맨 마지막에 와야 한다.


<img src= "https://github.com/happiipark/Java/assets/124240328/c77c1515-ce34-4536-96ee-21f4bafc6baa" width="800" height ="400" />

### 예외의 발생과 catch블럭

- try블럭에서 예외가 발생하면, 발생항 예외를 처리할 catch블럭을 찾는다
- 첫번째 catch블럭부터 순서대로 찾아 내려가며, 일치하는 catch블럭이 없으면 예외는 처리되지 않는다.
- 예외의 최고 조상인 Exception을 처리하는 catch블럭은 모든 종류의 예외를 처리할 수 있다

(**반드시 마지막 cathc블럭이어야 한다.**)
- 발생한 예외 객체를 catch블럭의 참조변수로 접근할 수 있다

```
ex)
try {
  tot += arr1[i] / arr2[i];
} catch (Exception e) { (1)
  count--;
} catch (ArithmeticException e) { (2)
  count--;
} catch (ArrayIndexOutOfBoundsException e) { (3)
  count--;
} 

위와 같이 작성할 경우 캐치문 (2), (3)은 에러가 발생하게 된다. 따라서

try {
  tot += arr1[i] / arr2[i];
} catch (ArithmeticException e) {
  count--;
} catch (ArrayIndexOutOfBoundsException e) {
  count--;
} catch (Exception e) {
  count--;
}
			
위처럼 변경해 줄 경우 에러 발생 X
```

```
printStackTrace() - 예외발생 당시의 호출스택(Call Stack)에 있엇던 메서드의 정보와 예외 메시지를 화면에 출력한다.
getMessage() - 발생한 예외클래스의 인스턴스에 저장된 메시지를 얻을 수 있다.
```

### finally블럭

- 예외의 발생여부와 관게없이 실행되어야 하는 코드를 넣는다.
- 선택적으로 사용할 수 있으며, try-catch-finally의 순서로 실행되고 예외 미발생시, try->finally의 순서로 실행된다
- try 또는 catch블럭에서 return문을 만나도 fianlly 블럭은 수행된다.

```
try{
  // 예외가 발생할 가능성이 있는 문장 작성
} catch (Exception e){
  // 예외 처리를 위한 문장 작성
} finally {
  // 예외의 발생여부에 관계없이 항상 수행되어야하는 문장 작성
  // finally 블럭은 try-catch문의 맨 마지막에 위치해야 함.
}
```

* finally 사용 이유
* try와 catch 안에 return이 있을경우 메서드 사용이 종료되니까
* finally를 사용할 경우 return을 만나도 finally까지 다 실행되고 종료
* --> 예외 발생 상관없이 

### 메서드에 예외 선언하기

- 예외를 처리하는 또 다른 방법
- 사실은 예외를 처리하는 것이 아니라, 예외를 호출한 메서드로 전달해주는 것
- 호출한 메서드에서 예외처리를 해야만 할 때 사용

```
void method() throws Exception1, Exception2, ..., ExceptionN {
}  // 메서드의 내용
```

```
public class ExceptionTest4 {
	
	public static void method1() throws Exception{  // throws를 통해 메소드 호출 부분에 예외를 던지기
		String str = null;
		try {
		System.out.println(str.toString());
		} catch (NullPointerException e) {
			System.out.println("method1에서 처리");
			throw new Exception("예외 다시 생성");
			// 런타임 exception이 아닌 예외를 발생 시킬 경우에는
			// throws Exception을 해줘야 한다 어디에? 메소드 선언부에!
		}
	}
	public static void main(String[] args) {
		try {
			method1();
		} catch(Exception e) {  // throws를 통해 받은 예외  catch문을 통하여 처리
			System.out.println("main에서 처리");
		}
		System.out.println("main 나머지 처리");
	}
}
// 꼭 에러가 난 메소드에서 예외처리를 안 해도 된다
// 메소드를 호출한 부분에서 처리 가능하다.
```


- 메인에서는 throws를 하지 않는다 
- 왜냐하면 main은 에러를 넘겨줄 곳이 없기 때문에 죽는다(강제로 죽이기)

* runtime exception은 런타임 도중 생기는 에러라 코드가 실행되지만
* Exception은 강제로 발생시킨 에러라 처리를 해줘야만 코드가 실행된다.


### 예외 되던지기 (re-throwing)

- 예외를 처리한 후에 다시 예외를 생성해서 호출한 메서드로 전달하는 것
- 예외가 발생한 메서드와 호출한 메서드, 양쪽에서 예외를 처리해야 하는 경우에 사용


### 사용자정의 예외 만들기

- 기존의 예외 클래스를 상속받아서 새로운 예외 클래스를 정의할 수 있다.

```
class MyException extends Exception {
	MyException (String msg) { // 문자열을 매개변수로 받는 생성자
		super(msg); // 조상인 Exception클래스의 생성자를 호출
	}
}
```

- 에러코드를 저장할 수 있게 ERR_CODE와 getErrCode()를 멤버로 추가

```
class MyException extends Exception{
	// 에러 코드 값을 저장하기 위한 필드 추가
	private final int ERR_CODE;
	
	MyException (String msg, int errCode) {  // 생성자
		super(msg);
		ERR_CODE = errCode;
	}
	
	MyException (String msg) { // 생성자
		this(msg,100); // ERR_CODE를 100(기본값)으로 초기화
	}
	
	public int getErrCode() { // 에러 코드를 얻을 수 있는 메서드 추가
		return ERR_CODE; 	  // 이 메서드는 주로 getMessage()와 함께 사용예상
	}
}
```


---


## Object클래스

### Object클래스의 메서드

- 모든 클래스의 최고 조상. 오직 11개의 메서드만을 가지고 있음!
- equals(), hashCode(), toString()은 적절히 오버라이딩해야 한다.
- notify(), wait() 등은 쓰레드와 관련된 메서드

* toString() : System.out.println()에 객체를 담게되면 자동으로 toString() 함수를 실행하게 된다!

### equals(Object obj)

- 객체 자신과 주어진 객체(obj)를 비교. 같으면 true, 다르면 fasle
- Object클래스에 정의된 equals()는 참조변수 값(객체의 주소)을 비교

```
public boolean equals(Object obj) {
	return (this == obj);
}
```

- equals()를 오버라이딩해서 멤버변수의 값을 비교하도록 바꾼다

```
class Person {
	long id;
	
	public boolean equals(Object obj) {
		if(obj != null && obj instanceof Person){
			return id == ((Person)obj).id;
		} else {
			return false;
		}
	}
	
	Person (long id){
		this.id = id;
	}
}
```

```
Ex)
class Person {
	String name;
	int age;
	Person(String name, int age){
		this.name = name;
		this.age = age;
	}
	
	@Override
	public String toString() {
		return "이름 : "+ name + ", 나이 : " + age;
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj instanceof Person == false) {
			return false;
		}
		Person p = (Person)obj;
		return name.equals(p.name) && age == p.age;
	}
	
}
public class ToStringTest {
	public static void main(String[] args) {
		Person p1 = new Person("hong",30);
		Person p2 = new Person("hong",30);
//		String str = "!!!+p";
		System.out.println(p1==p2);
		System.out.println(p1.equals(p2));
	}
}
```

예제를 따라하며 이해 안 갔던 부분 

- return name.equals(p.name) && age == p.age;
- 해당 부분에서 name은 어떻게 값을 가져오나 어떤 값을 가져오나 이해가 안 가서 궁금했는데
- 생각해보니 equals 메소드를 p1을 통하여 호출하니 당연히 p1의 name 값을 가져오는게 이해가 되었다
